rest_framework/serializers.py

```diff
@@ -159,7 +161,21 @@ def is_valid(self, raise_exception=False):
     def data(self):
         if not hasattr(self, '_data'):
             if self.instance is not None and not getattr(self, '_errors', None):
-                self._data = self.to_representation(self.instance)
+                try:
+                    self._data = self.to_representation(self.instance)
+                except AttributeError as exc:
+                    if isinstance(self.instance, (list, tuple, QuerySet)):
+                        msg = (
+                            'The reason for this exception might be that you '
+                            'have passed a %s as `instance` argument to the '
+                            'serializer but did not set `many=True`.'
+                            % type(self.instance).__name__)
+                        six.reraise(
+                            type(exc),
+                            type(exc)(str(exc) + '. ' + msg),
+                            sys.exc_info()[2])
+                    else:
+                        six.reraise(*sys.exc_info())
```

Comment:
I'd be happier with this if we weren't using the (to some folks) slightly obscure `exc_info` etc. Can we instead just use something like...

    except AttributeError as exc:
        if isinstance(self.instance, (list, tuple, QuerySet)):
            # If it appears that this error was due to the user omitting `many=True` then
            # we raise a more helpful error message.
            msg = "..." # Possibly include original exception text as per some of the try..catch blocks on model serializer.
            raise AttributeError(msg)
        raise